### PR TITLE
[flang][cuda] Add bind c to cudadevice procedures

### PR DESCRIPTION
**Pull request by clementval**
This patch adds bind c names to functions and subroutines in cudadevice so they can be lowered and not hit the intrinsic procedure TODOs.

### CODE REVIEW

**Comment by clementval** [comment]
Gentle ping

File: flang/module/cudadevice.f90 [hunk 1]
```diff
@@ -18,34 +18,34 @@ module cudadevice
...
-    attributes(device) integer function syncthreads_or(value)
+    attributes(device) integer function syncthreads_or(value) bind(c, name='__syncthreads_or')
       integer :: value
     end function
   end interface
   public :: syncthreads_or
 
   interface
-    attributes(device) subroutine syncwarp(mask)
+    attributes(device) subroutine syncwarp(mask) bind(c, name='__syncwrap')
```
Comment (vzakhari) [inline flang/module/cudadevice.f90:48]:
```suggestion
    attributes(device) subroutine syncwarp(mask) bind(c, name='__syncwarp')
```

**Review by vzakhari — APPROVED** [review]
LGTM, except one typo.

File: flang/module/cudadevice.f90 (hunk 1 above)
Comment (clementval) [reply flang/module/cudadevice.f90]: Wow! Nice catch!

**Review by wangzpgi — APPROVED** [review]

File: flang/test/Lower/CUDA/cuda-device-proc.cuf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+! CHECK: fir.call @__syncwrap(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> ()
+! CHECK: fir.call @__threadfence()
+! CHECK: fir.call @__threadfence_block()
+! CHECK: fir.call @__threadfence_system()
+! CHECK: %{{.*}} = fir.call @__syncthreads_and(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> i32
+! CHECK: %{{.*}} = fir.call @__syncthreads_count(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> i32
+! CHECK: %{{.*}} = fir.call @__syncthreads_or(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> i32
+
+! CHECK: func.func private @__syncthreads() attributes {cuf.proc_attr = #cuf.cuda_proc<device>, fir.bindc_name = "__syncthreads"}
+! CHECK: func.func private @__syncwrap(!fir.ref<i32> {cuf.data_attr = #cuf.cuda<device>}) attributes {cuf.proc_attr = #cuf.cuda_proc<device>, fir.bindc_name = "__syncwrap"}
```
Comment (clementval) [inline flang/test/Lower/CUDA/cuda-device-proc.cuf:30]:
```suggestion
! CHECK: func.func private @__syncwarp(!fir.ref<i32> {cuf.data_attr = #cuf.cuda<device>}) attributes {cuf.proc_attr = #cuf.cuda_proc<device>, fir.bindc_name = "__syncwarp"}
```
```suggestion
! CHECK: func.func private @__syncwrap(!fir.ref<i32> {cuf.data_attr = #cuf.cuda<device>}) attributes {cuf.proc_attr = #cuf.cuda_proc<device>, fir.bindc_name = "__syncwrap"}
```

File: flang/test/Lower/CUDA/cuda-device-proc.cuf [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  call threadfence_block()
+  call threadfence_system()
+  ret = syncthreads_and(1)
+  ret = syncthreads_count(1)
+  ret = syncthreads_or(1)
+end
+
+! CHECK-LABEL: func.func @_QPdevsub() attributes {cuf.proc_attr = #cuf.cuda_proc<global>}
+! CHECK: fir.call @__syncthreads()
+! CHECK: fir.call @__syncwrap(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> ()
```
Comment (clementval) [inline flang/test/Lower/CUDA/cuda-device-proc.cuf:21]:
```suggestion
! CHECK: fir.call @__syncwarp(%{{.*}}) fastmath<contract> : (!fir.ref<i32>) -> ()
```

**Pull request merged**